cli/src/main/scala/com/codacy/analysis/cli/formatter/Sarif.scala

```diff
@@ -53,53 +52,55 @@ private[formatter] class Sarif(val stream: PrintStream, val executionDirectory:
-            version = toolSpec.version.fold("0.0.0-unknown")(_.value.stripPrefix("v")),
-            rules = rules))
+      val driver = SarifReport.Tool(
+        SarifReport.Driver(
+          name = s"${toolSpec.name.value.capitalize} (reported by Codacy)",
+          version = toolSpec.version.fold("0.0.0-unknown")(_.value.stripPrefix("v")),
+          rules = securityRules ++ nonSecurityRules))
 
-        val invocations =
-          List(
-            SarifReport.Invocation(
-              executionSuccessful = true,
-              workingDirectory = SarifReport.ArtifactLocation(uri = rootDirectory)))
+      val invocations =
+        List(SarifReport
+          .Invocation(executionSuccessful = true, workingDirectory = SarifReport.ArtifactLocation(uri = rootDirectory)))
 
-        val artifacts: List[SarifReport.Artifact] = {
-          val filenames: Set[String] = securityIssues.map(_.filename.toString)(collection.breakOut)
-          filenames.map(filename => SarifReport.Artifact(SarifReport.ArtifactLocation(s"$rootDirectory/$filename")))(
-            collection.breakOut)
-        }
+      val artifacts = createArtifacts(securityIssues ++ nonSecurityIssues)
 
-        val results: List[SarifReport.Result] = securityIssues
-          .groupBy(_.filename)
-          .flatMap { case (filePath, issues) => createResults(filePath, issues, artifacts, rules) }(collection.breakOut)
+      val securityResults = createResults(securityIssues, artifacts, securityRules, securityIssues = true)
+      val nonSecurityResults =
+        createResults(nonSecurityIssues, artifacts, nonSecurityRules, securityIssues = false)
 
-        runs.add(SarifReport.Run(tool = driver, results = results, invocations = invocations, artifacts = artifacts))
+      runs.add(
+        SarifReport.Run(
+          tool = driver,
+          results = securityResults ++ nonSecurityResults,
+          invocations = invocations,
+          artifacts = artifacts))
 
-        ()
-
-      case _ => // Do nothing
+      ()
     }
   }
 
-  private def filterSecurityIssues(toolSpec: com.codacy.plugins.api.results.Tool.Specification,
-                                   analysisResults: Seq[Result]): Seq[Issue] = {
+  private def categorizeIssues(toolSpec: com.codacy.plugins.api.results.Tool.Specification,
+                               analysisResults: Seq[Result]): (Seq[Issue], Seq[Issue]) = {
     // HACK: Seems like the issues (`issue.category`) do not have the right category
     //   while in the specification (`toolSpec.patterns[].category`) the pattern has the right category
     val patternsCategoryMap: Map[String, Pattern.Category] =
       toolSpec.patterns.map(pattern => (pattern.patternId.value, pattern.category))(collection.breakOut)
-    analysisResults.collect {
-      case issue: Issue if patternsCategoryMap.get(issue.patternId.value).contains(Pattern.Category.Security) =>
-        issue
+
+    val analysisIssues = analysisResults.collect {
```

Comment:
What do you think about doing all this logic in a foldLeft?
I think probably the code would be very clear and smaller since would require less transformations.